### PR TITLE
support operator output with multiple types

### PR DESCRIPTION
**Pull request by mwdchang**
### Summary
Initial work to support operator output that can contain more than one type. For example calibrate run create a dataset + config. The way this is handled is to piggy back on the current notion where we "accept" one or more types.

```
type: 'datasetId | modelConfigId`
```

and the value of the output being an object keyed by the type:
```
value = [{
  datasetId: 'abcdefg',
  modelConfigId: 'aaaaabbbbbb'
}]
```

I largely removed the "allowMultiple" idea, as it isn't been used from what I can gather and it creates a boatload of edge cases that we would need to deal with.

### Testing
In unit tests
```
yarn workspace hmi-client run test workflow
```

### CODE REVIEW

File: packages/client/hmi-client/src/services/workflow.ts [hunk 1]
```diff
@@ -135,26 +167,38 @@ export const addEdge = (
...
 		(!targetInputPort.acceptMultiple && targetInputPort.status === WorkflowPortStatus.CONNECTED)
 	) {
 		return;
 	}
 
 	// Transfer data value/reference
-	if (targetInputPort.acceptMultiple && targetInputPort.value && sourceOutputPort.value) {
-		targetInputPort.label = `${sourceOutputPort.label},${targetInputPort.label}`;
-		targetInputPort.value = [...targetInputPort.value, ...sourceOutputPort.value];
+	targetInputPort.label = sourceOutputPort.label;
```
Comment (mwdchang) [inline packages/client/hmi-client/src/services/workflow.ts:189]: The port labelling part will likely need a second pass, not exactly sure what we want to show if a port has multiple types.

File: packages/client/hmi-client/src/services/workflow.ts [hunk 2]
```diff
@@ -135,26 +167,38 @@ export const addEdge = (
 	if (existingEdge) return;
 
 	// Check if type is compatible
-	const allowedTypes = targetInputPort.type.split('|');
+	const outputTypes = sourceOutputPort.type.split('|');
+	const allowedInputTypes = targetInputPort.type.split('|');
```
Comment (YohannParis) [inline packages/client/hmi-client/src/services/workflow.ts:171]: Should those be trimmed of whitespaces?

File: packages/client/hmi-client/src/services/workflow.ts [hunk 3]
```diff
@@ -135,26 +167,38 @@ export const addEdge = (
...
+	const intersectionTypes = _.intersection(outputTypes, allowedInputTypes);
+
+	// Not supported if there are more than oen matches
```
Comment (YohannParis) [inline packages/client/hmi-client/src/services/workflow.ts:174]:
```suggestion
	// Not supported if there are more than one matche
```

File: packages/client/hmi-client/src/services/workflow.ts (hunk 1 above)
Comment (YohannParis) [reply packages/client/hmi-client/src/services/workflow.ts:189]: Maybe a simple ` or `.

**Review by YohannParis — APPROVED** [review]
LGTM

File: packages/client/hmi-client/src/services/workflow.ts (hunk 2 above)
Comment (mwdchang) [reply packages/client/hmi-client/src/services/workflow.ts:171]: yep, thx, will fix

File: packages/client/hmi-client/src/services/workflow.ts [hunk 4]
```diff
@@ -108,6 +108,38 @@ export const addNode = (
 	wf.nodes.push(node);
 };
 
+/**
+ * Create an eddge between two nodes, and transfer the value from
```
Comment (shawnyama) [inline packages/client/hmi-client/src/services/workflow.ts:112]:
```suggestion
* Create an edge between two nodes, and transfer the value from
```

File: packages/client/hmi-client/src/services/workflow.ts (hunk 1 above)
Comment (shawnyama) [reply packages/client/hmi-client/src/services/workflow.ts:189]: There's a `getPortLabel` function in this file that does this

File: packages/client/hmi-client/src/services/workflow.ts [hunk 5]
```diff
@@ -108,6 +108,38 @@ export const addNode = (
...
+ *
+ * The target-input can accept multiple types, but this is more of an exception
+ * rather than the norm. This is resolved by checking one of the accepted type
+ * matches the provided type.
+ *
+ * The second case is when the source produce multiple types, a similar check, but
+ * on the reverse side is performed.
+ *
+ *
+ * Ambiguity arises when the source provide multiple types and the target accepts multiple
```
Comment (shawnyama) [inline packages/client/hmi-client/src/services/workflow.ts:136]:
```suggestion
* Ambiguity arises when the source provides multiple types and the target accepts multiple
```

File: packages/client/hmi-client/src/services/workflow.ts [hunk 6]
```diff
@@ -135,26 +167,38 @@ export const addEdge = (
 	if (existingEdge) return;
 
 	// Check if type is compatible
-	const allowedTypes = targetInputPort.type.split('|');
+	const outputTypes = sourceOutputPort.type.split('|');
+	const allowedInputTypes = targetInputPort.type.split('|');
+	const intersectionTypes = _.intersection(outputTypes, allowedInputTypes);
+
+	// Not supported if there are more than one matche
```
Comment (shawnyama) [inline packages/client/hmi-client/src/services/workflow.ts:174]:
```suggestion
  // Not supported if there are more than one match
```

**Review by shawnyama — APPROVED** [review]

File: packages/client/hmi-client/src/services/workflow.ts (hunk 1 above)
Comment (mwdchang) [reply packages/client/hmi-client/src/services/workflow.ts:189]: getPortLabel happens "after" we set the port label if I recall - this here happens at a lower level.
